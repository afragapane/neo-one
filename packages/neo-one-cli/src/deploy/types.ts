import {
  ForwardOptions,
  GetOptions,
  Param,
  TransactionOptions,
  UserAccountID,
  Transfer,
  Hash256String,
} from '@neo-one/client-common';
import { Deferred } from './createDeferred';

export type MigrationContractsBase = object;

export type Migration<Contracts extends MigrationContractsBase> = (
  contracts: Contracts,
  network: string,
  userAccountIDs: readonly UserAccountID[],
) => void;

export type MigrationParam = Param | Promise<Param>;

export interface Action {
  readonly contract: string;
  readonly method: string;
  readonly params: readonly MigrationParam[];
  readonly options?: TransactionOptions & GetOptions;
  readonly forwardOptions?: ForwardOptions<any>;
  readonly transfer?: Transfer;
  readonly hash?: Hash256String;
  readonly deferred: Deferred;
}
