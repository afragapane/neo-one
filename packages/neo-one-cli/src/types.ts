import yargs from 'yargs';

// tslint:disable-next-line no-any
export type Yarguments<Argv extends yargs.Argv<any>> = yargs.Arguments<
  Argv extends yargs.Argv<infer Value> ? Value : never
>;

export interface Command {
  readonly bin: string;
  readonly args: readonly string[];
}
