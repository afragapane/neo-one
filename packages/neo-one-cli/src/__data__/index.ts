export * from './getClients';
export * from './getContracts';
