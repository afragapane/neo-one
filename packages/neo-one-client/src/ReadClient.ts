import { ScriptBuilderParam } from '@neo-one/client-core';
import { Monitor } from '@neo-one/monitor';
import * as args from './args';
import { createReadSmartContract } from './sc';
import {
  Account,
  AddressString,
  Asset,
  Block,
  BlockFilter,
  BufferString,
  Contract,
  DataProvider,
  GetOptions,
  Hash256String,
  Input,
  Output,
  Peer,
  RawAction,
  RawCallReceipt,
  ReadSmartContract,
  ReadSmartContractAny,
  ReadSmartContractDefinition,
  StorageItem,
  Transaction,
} from './types';

export class ReadClient<TDataProvider extends DataProvider = DataProvider> {
  public readonly dataProvider: TDataProvider;

  public constructor(dataProvider: TDataProvider) {
    this.dataProvider = dataProvider;
  }

  public async getAccount(address: AddressString, monitor?: Monitor): Promise<Account> {
    return this.dataProvider.getAccount(args.assertAddress('address', address), monitor);
  }

  public async getAsset(hash: Hash256String, monitor?: Monitor): Promise<Asset> {
    return this.dataProvider.getAsset(args.assertHash256('hash', hash), monitor);
  }

  public async getBlock(hash: number | Hash256String, optionsIn?: GetOptions): Promise<Block> {
    const options = args.assertGetOptions('options', optionsIn);
    if (typeof hash === 'number') {
      return this.dataProvider.getBlock(hash, options);
    }

    return this.dataProvider.getBlock(args.assertHash256('hash', hash), options);
  }

  public iterBlocks(filter?: BlockFilter): AsyncIterable<Block> {
    return this.dataProvider.iterBlocks(args.assertBlockFilter('filter', filter));
  }

  public async getBestBlockHash(monitor?: Monitor): Promise<Hash256String> {
    return this.dataProvider.getBestBlockHash(monitor);
  }

  public async getBlockCount(monitor?: Monitor): Promise<number> {
    return this.dataProvider.getBlockCount(monitor);
  }

  public async getContract(address: AddressString, monitor?: Monitor): Promise<Contract> {
    return this.dataProvider.getContract(args.assertAddress('address', address), monitor);
  }

  public async getMemPool(monitor?: Monitor): Promise<ReadonlyArray<Hash256String>> {
    return this.dataProvider.getMemPool(monitor);
  }

  public async getTransaction(hash: Hash256String, monitor?: Monitor): Promise<Transaction> {
    return this.dataProvider.getTransaction(args.assertHash256('hash', hash), monitor);
  }

  public async getOutput(input: Input, monitor?: Monitor): Promise<Output> {
    return this.dataProvider.getOutput(input, monitor);
  }

  public async getConnectedPeers(monitor?: Monitor): Promise<ReadonlyArray<Peer>> {
    return this.dataProvider.getConnectedPeers(monitor);
  }

  // tslint:disable-next-line no-any
  public smartContract<T extends ReadSmartContract<any> = ReadSmartContractAny>(
    definition: ReadSmartContractDefinition,
  ): T {
    return createReadSmartContract({
      definition: args.assertReadSmartContractDefinition('definition', definition),
      client: this,
      // tslint:disable-next-line no-any
    }) as any;
  }

  public async getStorage(address: AddressString, key: BufferString, monitor?: Monitor): Promise<StorageItem> {
    return this.dataProvider.getStorage(args.assertAddress('address', address), args.assertBuffer('key', key), monitor);
  }

  public iterStorage(address: AddressString, monitor?: Monitor): AsyncIterable<StorageItem> {
    return this.dataProvider.iterStorage(args.assertAddress('address', address), monitor);
  }

  public __iterActionsRaw(filter?: BlockFilter): AsyncIterable<RawAction> {
    return this.dataProvider.iterActionsRaw(args.assertBlockFilter('filter', filter));
  }

  public async __call(
    contract: AddressString,
    method: string,
    params: ReadonlyArray<ScriptBuilderParam | undefined>,
    monitor?: Monitor,
  ): Promise<RawCallReceipt> {
    return this.dataProvider.call(contract, method, params, monitor);
  }
}
