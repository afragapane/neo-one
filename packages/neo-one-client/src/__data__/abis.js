/* @flow */
import type {
  ABIEvent,
  ABIFunction,
  ABIParameter,
  ABIReturn,
  ABI,
} from '../types';

export const returns = {
  Signature: { type: 'Signature' },
  Boolean: { type: 'Boolean' },
  Hash160: { type: 'Hash160' },
  Hash256: { type: 'Hash256' },
  ByteArray: { type: 'ByteArray' },
  PublicKey: { type: 'PublicKey' },
  String: { type: 'String' },
  InteropInterface: { type: 'InteropInterface' },
  Void: { type: 'Void' },
  Integer: { type: 'Integer', decimals: 7 },
  Array: { type: 'Array', value: { type: 'String' } },
};

export const parameters = {
  Signature: { name: 'signatureName', type: 'Signature' },
  Boolean: { name: 'booleanName', type: 'Boolean' },
  Integer: { name: 'intName', type: 'Integer', decimals: 7 },
  Hash160: { name: 'hash160Name', type: 'Hash160' },
  Hash256: { name: 'hash256Name', type: 'Hash256' },
  ByteArray: { name: 'byteName', type: 'ByteArray' },
  PublicKey: { name: 'pubkeyName', type: 'PublicKey' },
  String: { name: 'stringName', type: 'String' },
  Array: {
    name: 'arrayName',
    type: 'Array',
    value: {
      name: 'stringName',
      type: 'String',
    },
  },
  InteropInterface: { name: 'interopName', type: 'InteropInterface' },
  Void: { name: 'voidName', type: 'Void' },
};

export const abiFunction = (
  constant: boolean = false,
  params: Array<ABIParameter> = [parameters.String],
  verify: boolean = false,
  name: string = 'funcName',
  returnType: ABIReturn = returns.Void,
): ABIFunction => ({
  name,
  constant,
  parameters: params,
  verify,
  returnType,
});

export const abiEvent = (
  name: string = 'eventName',
  params: Array<ABIParameter> = [parameters.String],
): ABIEvent => ({
  name,
  parameters: params,
});

export const abi = (
  functions: Array<ABIFunction> = [abiFunction()],
  events?: Array<ABIEvent> = [],
): ABI => ({
  functions,
  events,
});
