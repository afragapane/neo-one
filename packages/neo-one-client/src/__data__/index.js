/* @flow */
export { default as keys } from './keys';
export { default as transactions } from './transactions';
