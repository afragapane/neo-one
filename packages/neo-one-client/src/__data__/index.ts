export * from './keys';
export * from './data';
export * from './transactions';
export * from './factory';
