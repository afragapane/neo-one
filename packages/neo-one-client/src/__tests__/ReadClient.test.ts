import { AsyncIterableX } from '@reactivex/ix-es2015-cjs/asynciterable';
import { toArray } from '@reactivex/ix-es2015-cjs/asynciterable/toarray';
import { data, factory, keys } from '../__data__';
import { ReadClient } from '../ReadClient';
import { DataProvider } from '../types';

describe('ReadClient', () => {
  const network = 'main';
  const getAccount = jest.fn();
  const getAsset = jest.fn();
  const getBlock = jest.fn();
  const iterBlocks = jest.fn();
  const getBestBlockHash = jest.fn();
  const getBlockCount = jest.fn();
  const getContract = jest.fn();
  const getMemPool = jest.fn();
  const getTransaction = jest.fn();
  const getOutput = jest.fn();
  const getConnectedPeers = jest.fn();
  const getStorage = jest.fn();
  const iterStorage = jest.fn();
  const iterActionsRaw = jest.fn();
  const call = jest.fn();
  const provider: DataProvider = {
    network,
    getAccount,
    getAsset,
    getBlock,
    iterBlocks,
    getBestBlockHash,
    getBlockCount,
    getContract,
    getMemPool,
    getTransaction,
    getOutput,
    getConnectedPeers,
    getStorage,
    iterStorage,
    iterActionsRaw,
    call,
  };
  let client: ReadClient;
  beforeEach(() => {
    client = new ReadClient(provider);
  });

  test('getAccount', async () => {
    const value = factory.createAccount();
    getAccount.mockImplementationOnce(() => value);

    const result = await client.getAccount(keys[0].address);

    expect(result).toEqual(value);
    expect(getAccount.mock.calls).toMatchSnapshot();
  });

  test('getAsset', async () => {
    const value = factory.createAsset();
    getAsset.mockImplementationOnce(() => value);

    const result = await client.getAsset(data.hash256s.a);

    expect(result).toEqual(value);
    expect(getAsset.mock.calls).toMatchSnapshot();
  });

  test('getBlock', async () => {
    const value = factory.createBlock();
    getBlock.mockReset();
    getBlock.mockImplementationOnce(() => value);

    const result = await client.getBlock(data.hash256s.a);

    expect(result).toEqual(value);
    expect(getBlock.mock.calls).toMatchSnapshot();
  });

  test('getBlock - number', async () => {
    const value = factory.createBlock();
    getBlock.mockReset();
    getBlock.mockImplementationOnce(() => value);

    const result = await client.getBlock(0);

    expect(result).toEqual(value);
    expect(getBlock.mock.calls).toMatchSnapshot();
  });

  test('iterBlocks', async () => {
    const value = factory.createBlock();
    iterBlocks.mockImplementationOnce(() => AsyncIterableX.from([value]));

    const result = await toArray(client.iterBlocks({ indexStart: 3, indexStop: 4 }));

    expect(result).toEqual([value]);
    expect(iterBlocks.mock.calls).toMatchSnapshot();
  });

  test('getBestBlockHash', async () => {
    const value = data.hash256s.a;
    getBestBlockHash.mockImplementationOnce(() => value);

    const result = await client.getBestBlockHash();

    expect(result).toEqual(value);
    expect(getBestBlockHash.mock.calls).toMatchSnapshot();
  });

  test('getBlockCount', async () => {
    const value = 10;
    getBlockCount.mockImplementationOnce(() => value);

    const result = await client.getBlockCount();

    expect(result).toEqual(value);
    expect(getBlockCount.mock.calls).toMatchSnapshot();
  });

  test('getContract', async () => {
    const value = factory.createContract();
    getContract.mockImplementationOnce(() => value);

    const result = await client.getContract(keys[0].address);

    expect(result).toEqual(value);
    expect(getContract.mock.calls).toMatchSnapshot();
  });

  test('getMemPool', async () => {
    const value = [data.hash256s.a];
    getMemPool.mockImplementationOnce(() => value);

    const result = await client.getMemPool();

    expect(result).toEqual(value);
    expect(getMemPool.mock.calls).toMatchSnapshot();
  });

  test('getTransaction', async () => {
    const value = factory.createContractTransaction();
    getTransaction.mockImplementationOnce(() => value);

    const result = await client.getTransaction(data.hash256s.a);

    expect(result).toEqual(value);
    expect(getTransaction.mock.calls).toMatchSnapshot();
  });

  test('getOutput', async () => {
    const value = factory.createOutput();
    getOutput.mockImplementationOnce(() => value);

    const result = await client.getOutput(factory.createInput());

    expect(result).toEqual(value);
    expect(getOutput.mock.calls).toMatchSnapshot();
  });

  test('getConnectedPeers', async () => {
    const value = [factory.createPeerJSON()];
    getConnectedPeers.mockImplementationOnce(() => value);

    const result = await client.getConnectedPeers();

    expect(result).toEqual(value);
    expect(getConnectedPeers.mock.calls).toMatchSnapshot();
  });

  test('smartContract', () => {
    const result = client.smartContract(factory.createReadSmartContractDefinition());

    expect(result).toMatchSnapshot();
  });

  test('getStorage', async () => {
    const value = factory.createStorageItem();
    getStorage.mockImplementationOnce(() => value);

    const result = await client.getStorage(keys[0].address, data.buffers.a);

    expect(result).toEqual(value);
    expect(getStorage.mock.calls).toMatchSnapshot();
  });

  test('iterStorage', async () => {
    const value = [factory.createStorageItem()];
    iterStorage.mockImplementationOnce(() => AsyncIterableX.from(value));

    const result = await toArray(client.iterStorage(keys[0].address));

    expect(result).toEqual(value);
    expect(getStorage.mock.calls).toMatchSnapshot();
  });

  test('__iterActionsRaw', async () => {
    const value = [factory.createRawNotification()];
    iterActionsRaw.mockImplementationOnce(() => AsyncIterableX.from(value));

    const result = await toArray(client.__iterActionsRaw());

    expect(result).toEqual(value);
    expect(iterActionsRaw.mock.calls).toMatchSnapshot();
  });

  test('__call', async () => {
    const value = factory.createRawCallReceipt();
    call.mockImplementationOnce(() => value);

    const result = await client.__call(keys[0].address, 'deploy', []);

    expect(result).toEqual(value);
    expect(iterActionsRaw.mock.calls).toMatchSnapshot();
  });
});
