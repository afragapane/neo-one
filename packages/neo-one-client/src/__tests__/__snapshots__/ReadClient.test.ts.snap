// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadClient __call 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient __iterActionsRaw 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient getAccount 1`] = `
Array [
  Array [
    "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    undefined,
  ],
]
`;

exports[`ReadClient getAsset 1`] = `
Array [
  Array [
    "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
    undefined,
  ],
]
`;

exports[`ReadClient getBestBlockHash 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient getBlock - number 1`] = `
Array [
  Array [
    0,
    undefined,
  ],
]
`;

exports[`ReadClient getBlock 1`] = `
Array [
  Array [
    "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
    undefined,
  ],
]
`;

exports[`ReadClient getBlockCount 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient getConnectedPeers 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient getContract 1`] = `
Array [
  Array [
    "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    undefined,
  ],
]
`;

exports[`ReadClient getMemPool 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient getOutput 1`] = `
Array [
  Array [
    Object {
      "hash": "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
      "index": 0,
    },
    undefined,
  ],
]
`;

exports[`ReadClient getStorage 1`] = `
Array [
  Array [
    "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    "b500",
    undefined,
  ],
]
`;

exports[`ReadClient getTransaction 1`] = `
Array [
  Array [
    "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
    undefined,
  ],
]
`;

exports[`ReadClient iterBlocks 1`] = `
Array [
  Array [
    Object {
      "indexStart": 3,
      "indexStop": 4,
    },
  ],
]
`;

exports[`ReadClient iterStorage 1`] = `
Array [
  Array [
    "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    "b500",
    undefined,
  ],
]
`;

exports[`ReadClient smartContract 1`] = `
Object {
  "balanceOf": [Function],
  "convertAction": [Function],
  "decimals": [Function],
  "definition": Object {
    "abi": Object {
      "events": Array [
        Object {
          "name": "transfer",
          "parameters": Array [
            Object {
              "default": undefined,
              "forwardedValue": undefined,
              "name": "from",
              "optional": true,
              "rest": undefined,
              "type": "Address",
            },
            Object {
              "default": undefined,
              "forwardedValue": undefined,
              "name": "to",
              "optional": true,
              "rest": undefined,
              "type": "Address",
            },
            Object {
              "decimals": 8,
              "default": undefined,
              "forwardedValue": undefined,
              "name": "amount",
              "optional": undefined,
              "type": "Integer",
            },
          ],
        },
      ],
      "functions": Array [
        Object {
          "claim": undefined,
          "completeSend": undefined,
          "constant": true,
          "name": "name",
          "parameters": Array [],
          "receive": undefined,
          "refundAssets": undefined,
          "returnType": Object {
            "forwardedValue": undefined,
            "optional": undefined,
            "type": "String",
          },
          "send": undefined,
          "sendUnsafe": undefined,
        },
        Object {
          "claim": undefined,
          "completeSend": undefined,
          "constant": true,
          "name": "symbol",
          "parameters": Array [],
          "receive": undefined,
          "refundAssets": undefined,
          "returnType": Object {
            "forwardedValue": undefined,
            "optional": undefined,
            "type": "String",
          },
          "send": undefined,
          "sendUnsafe": undefined,
        },
        Object {
          "claim": undefined,
          "completeSend": undefined,
          "constant": true,
          "name": "decimals",
          "parameters": Array [],
          "receive": undefined,
          "refundAssets": undefined,
          "returnType": Object {
            "decimals": 0,
            "forwardedValue": undefined,
            "optional": undefined,
            "type": "Integer",
          },
          "send": undefined,
          "sendUnsafe": undefined,
        },
        Object {
          "claim": undefined,
          "completeSend": undefined,
          "constant": true,
          "name": "totalSupply",
          "parameters": Array [],
          "receive": undefined,
          "refundAssets": undefined,
          "returnType": Object {
            "decimals": 8,
            "forwardedValue": undefined,
            "optional": undefined,
            "type": "Integer",
          },
          "send": undefined,
          "sendUnsafe": undefined,
        },
        Object {
          "claim": undefined,
          "completeSend": undefined,
          "constant": undefined,
          "name": "transfer",
          "parameters": Array [
            Object {
              "default": undefined,
              "forwardedValue": undefined,
              "name": "from",
              "optional": undefined,
              "rest": undefined,
              "type": "Address",
            },
            Object {
              "default": undefined,
              "forwardedValue": undefined,
              "name": "to",
              "optional": undefined,
              "rest": undefined,
              "type": "Address",
            },
            Object {
              "decimals": 8,
              "default": undefined,
              "forwardedValue": undefined,
              "name": "value",
              "optional": undefined,
              "type": "Integer",
            },
          ],
          "receive": undefined,
          "refundAssets": undefined,
          "returnType": Object {
            "forwardedValue": undefined,
            "optional": undefined,
            "type": "Boolean",
          },
          "send": undefined,
          "sendUnsafe": undefined,
        },
        Object {
          "claim": undefined,
          "completeSend": undefined,
          "constant": true,
          "name": "balanceOf",
          "parameters": Array [
            Object {
              "default": undefined,
              "forwardedValue": undefined,
              "name": "account",
              "optional": undefined,
              "rest": undefined,
              "type": "Address",
            },
          ],
          "receive": undefined,
          "refundAssets": undefined,
          "returnType": Object {
            "decimals": 8,
            "forwardedValue": undefined,
            "optional": undefined,
            "type": "Integer",
          },
          "send": undefined,
          "sendUnsafe": undefined,
        },
      ],
    },
    "address": "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    "sourceMaps": undefined,
  },
  "iterActions": [Function],
  "iterEvents": [Function],
  "iterLogs": [Function],
  "iterStorage": [Function],
  "name": [Function],
  "symbol": [Function],
  "totalSupply": [Function],
}
`;
